torchgeo/datasets/seasonet.py

```diff
@@ -0,0 +1,456 @@
+        self.concat_seasons = concat_seasons
+        self.transforms = transforms
+        self.download = download
+        self.checksum = checksum
+        self.cmap = ListedColormap(np.array(self.colormap) / 255)
+
+        self._verify()
+
+        self.channels = 0
+        for b in bands:
+            self.channels += self.band_nums[b]
+
+        csv = pd.read_csv(os.path.join(self.root, "meta.csv"), index_col="Index")
+
+        if split is not None:
+            split_csv = pd.read_csv(
+                os.path.join(self.root, f"splits/{split}.csv"), header=None
+            )[0]
+            csv = csv.iloc[split_csv]
+
+        csv = csv[csv["Grid"].isin(grids)]
+        csv = csv[csv["Season"].isin(seasons)]
+        csv["Path"] = csv["Path"].apply(
+            lambda p: [os.path.join(self.root, p, os.path.basename(p))]
+        )
+
+        if self.concat_seasons > 1:
+            self.files = csv.groupby(["Latitude", "Longitude"])
+            self.files = self.files["Path"].agg(sum)
+            self.files = self.files[
+                self.files.apply(lambda d: len(d) >= self.concat_seasons)
+            ]
+        else:
+            self.files = csv["Path"]
+
+    def __getitem__(self, index: int) -> dict[str, Tensor]:
+        """Return an index within the dataset.
+
+        Args:
+            index: index to return
+
+        Returns:
+            sample at that index containing the image with shape SCxHxW,
+            where ``S = self.concat_seasons``, and the mask
+        """
+        image = self._load_image(index)
+        mask = self._load_target(index)
+        sample = {"image": image, "mask": mask}
+
+        if self.transforms is not None:
+            sample = self.transforms(sample)
+
+        return sample
+
+    def __len__(self) -> int:
+        """Return the number of data points in the dataset.
+
+        Returns:
+            length of the dataset
+        """
+        return len(self.files)
+
+    def _load_image(self, index: int) -> Tensor:
+        """Load image(s) for a single location.
+
+        Args:
+            index: index to return
+
+        Returns:
+            the stacked seasonal images
+        """
+        paths = self.files.iloc[index]
+        if self.concat_seasons > 1:
+            paths = random.sample(paths, self.concat_seasons)
+        tensor = torch.empty(self.concat_seasons * self.channels, *self.image_size)
+        for img_idx, path in enumerate(paths):
+            bnd_idx = 0
+            for band in self.bands:
+                with rasterio.open(f"{path}_{band}.tif") as f:
+                    array = f.read(
+                        out_shape=[f.count] + list(self.image_size),
+                        out_dtype="int32",
+                        resampling=Resampling.bilinear,
+                    )
+                image = torch.from_numpy(array).float()
+                c = img_idx * self.channels + bnd_idx
+                tensor[c : c + image.shape[0]] = image
```

Comment:
I usually just append to a list and concatenate at the end, not sure if that would be simpler

Reply:
That may be simpler but afaik this way should be a bit better performance-wise